hal/src/nRF52840/exflash_hal.cpp

```diff
@@ -0,0 +1,717 @@
+                /* Adjust addresses and size */
+                addr += size_aligned;
+                data_buf += size_aligned;
+                data_size -= size_aligned;
+            }
+        }
+    }
+
+    /* Read the remaining unaligned portion into a temporary buffer and copy into
+     * the destination buffer
+     */
+    if (data_size > 0) {
+        uint8_t tmpbuf[sizeof(uint32_t) * 2] __attribute__((aligned(4)));
+
+        /* Read the remainder */
+        const uintptr_t src_aligned = ADDR_ALIGN_WORD(addr);
+        const size_t size_aligned = ADDR_ALIGN_WORD_RIGHT((addr + data_size) - src_aligned);
+
+        ret = nrfx_qspi_read(tmpbuf, size_aligned, src_aligned);
+        CHECK_TRUE(ret == NRFX_SUCCESS, nrf_system_error(ret));
+
+        const unsigned offset_front = addr - src_aligned;
+
+        memcpy(data_buf, tmpbuf + offset_front, data_size);
+    }
+
+    return 0;
+}
+
+int hal_exflash_erase_sector(uintptr_t start_addr, size_t num_sectors) {
+    return erase_common(start_addr, num_sectors, NRF_QSPI_ERASE_LEN_4KB);
+}
+
+int hal_exflash_erase_block(uintptr_t start_addr, size_t num_blocks) {
+    return erase_common(start_addr, num_blocks, NRF_QSPI_ERASE_LEN_64KB);
+}
+
+int hal_exflash_copy_sector(uintptr_t src_addr, uintptr_t dest_addr, size_t data_size) {
+    FlashLock lk;
+
+    uint8_t data_buf[FLASH_OPERATION_TEMP_BLOCK_SIZE] __attribute__((aligned(4)));
+    unsigned index = 0;
+    uint16_t copy_len;
+    uint16_t sector_num;
+
+    if ((src_addr % sFLASH_PAGESIZE) || (dest_addr % sFLASH_PAGESIZE) || !(IS_WORD_ALIGNED(data_size))) {
+        return SYSTEM_ERROR_INVALID_ARGUMENT;
+    }
+
+    // erase sectors
+    sector_num = CEIL_DIV(data_size, sFLASH_PAGESIZE);
+    CHECK(hal_exflash_erase_sector(dest_addr, sector_num));
+
+    // memory copy
+    for (index = 0; index < data_size; index += copy_len) {
+        copy_len = std::min((data_size - index), sizeof(data_buf));
+        CHECK(hal_exflash_read(src_addr + index, data_buf, copy_len));
+        CHECK(hal_exflash_write(dest_addr + index, data_buf, copy_len));
+    }
+
+    return 0;
+}
+
+int hal_exflash_read_special(hal_exflash_special_sector_t sp, uintptr_t addr, uint8_t* data_buf, size_t data_size) {
+    const hal_exflash_params_t* flash_params = hal_exflash_get_params(flash_type);
+    CHECK_TRUE(flash_params, SYSTEM_ERROR_NOT_FOUND);
+
+    CHECK_TRUE(sp == HAL_EXFLASH_SPECIAL_SECTOR_OTP, SYSTEM_ERROR_INVALID_ARGUMENT);
+    CHECK_TRUE(addr + data_size <= flash_params->otp_size, SYSTEM_ERROR_OUT_OF_RANGE);
+    CHECK_TRUE(data_buf != nullptr, SYSTEM_ERROR_INVALID_ARGUMENT);
+    int ret = 0;
+
+    FlashLock lk;
+
+    if (flash_type == HAL_QSPI_FLASH_TYPE_GD25WQ64E) {
+        // Determine which OTP Register to start reading from
+        size_t otp_page = addr / GD25_SECURITY_REGISTER_SIZE;
+        size_t data_to_read = data_size;
+        size_t data_read = 0;
+        size_t read_amount = 0;
+
+        while (data_to_read && !ret) {
```

Comment:
Nit: move to some common code as it's reused below.

Reply:
Addressed this, let me know if refactor is ok